packages/vite/src/shared/runnerTransport.ts

```diff
@@ -0,0 +1,304 @@
+import { nanoid } from 'nanoid/non-secure'
+import type { CustomPayload, HotPayload } from 'types/hmrPayload'
+import { promiseWithResolvers } from './utils'
+
+export type RunnerTransportHandlers = {
+  onDisconnection: () => void
+}
+
+export type CreateRunnerTransport = (
+  handlers: RunnerTransportHandlers,
+) => RunnerTransportOptions
+
+/**
+ * "send and connect" or "invoke" must be implemented
+ */
+export interface RunnerTransportOptions {
+  connect?(handler: (data: HotPayload) => void): Promise<void> | void
+  disconnect?(): Promise<void> | void
+  send?(data: HotPayload): Promise<void> | void
+  invoke?(
+    data: HotPayload,
+  ): Promise<{ /** result */ r: any } | { /** error */ e: any }>
+  timeout?: number
+}
+
+type NormalizedRunnerTransportOptions = Omit<
+  RunnerTransportOptions,
+  'send' | 'invoke'
+> & {
+  connect?(
+    handler: ((data: HotPayload) => void) | undefined,
+  ): Promise<void> | void
+  send(data: HotPayload): void
+  invoke(name: string, data: any): any
+}
+
+const createInvokeableTransportOptions = (
+  transport: RunnerTransportOptions,
+): NormalizedRunnerTransportOptions => {
+  if (transport.invoke) {
+    const sendOrInvoke = transport.send ?? transport.invoke
+    return {
+      ...transport,
+      async send(data) {
+        await sendOrInvoke(data)
+      },
+      async invoke(name, data) {
+        const result = await transport.invoke!({
+          type: 'custom',
+          event: name,
+          invoke: 'send',
+          data,
+        } satisfies CustomPayload)
+        if ('e' in result) {
+          throw result.e
+        }
+        return result.r
+      },
+    }
+  }
+
+  if (!transport.send || !transport.connect) {
+    throw new Error(
+      'transport must implement send and connect when invoke is not implemented',
+    )
+  }
+
+  const rpcPromises = new Map<
+    string,
+    {
+      resolve: (data: any) => void
+      reject: (data: any) => void
+      timeoutId?: ReturnType<typeof setTimeout>
+    }
+  >()
+
+  return {
+    ...transport,
```

Comment:
My implementation was based on birpc at the time. Since timeout can be anything, I think we should also reject all hanging promises when the transport is closed if there are any (example: https://github.com/antfu/birpc/pull/21)

Reply:
Done in https://github.com/vitejs/vite/pull/18362/commits/3e5807a2e1497670993f3dcf08986a93e090f67e 👍